### PR TITLE
create watcher for eviction worker with timeout option

### PR DESCRIPTION
**Pull request by IgorEulalio**

### CODE REVIEW

File: actionners/kubernetes/drain/drain.go [hunk 1]
```diff
@@ -166,14 +181,24 @@ func Action(action *rules.Action, event *events.Event) (utils.LogLine, *model.Da
...
 
 	wg.Wait()
+	if config.WaitPeriod != 0 {
+		err = verifyEvictionHasFinished(client, config.WaitPeriod, nodeName, config)
+		if err != nil {
+			return utils.LogLine{
+				Objects: objects,
+				Error:   fmt.Sprintf("pods were not evited during the wait period of %v seconds for node %s.", config.WaitPeriod, nodeName),
+				Status:  "failure",
+			}, nil, fmt.Errorf("pods were not evited during the wait period of %v seconds for node %s", config.WaitPeriod, nodeName)
```
Comment (Issif) [inline actionners/kubernetes/drain/drain.go:199]:
```suggestion
			err = fmt.Errorf("pods were not evited during the wait period of %v seconds for node %s", config.WaitPeriod, nodeName)
			return utils.LogLine{
				Objects: objects,
				Error:   err.Error(),
				Status:  "failure",
			}, nil, err
```

File: utils/utils.go [hunk 2]
```diff
@@ -38,6 +38,10 @@ const (
 	colorStr string = "color"
 
 	ansiChars string = "[\u001B\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[a-zA-Z\\d]*)*)?\u0007)|(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PRZcf-ntqry=><~]))"
+
+	DaemonSetStr   = "DaemonSet"
+	StatefulSetStr = "StatefulSet"
+	ReplicaSetStr  = "ReplicaSet"
```
Comment (Issif) [inline utils/utils.go:44]: must be used in other actionners then (terminate, ...)

File: actionners/kubernetes/drain/drain.go [hunk 3]
```diff
@@ -189,6 +214,54 @@ func Action(action *rules.Action, event *events.Event) (utils.LogLine, *model.Da
...
+			}
+			pods, err := c.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
+				FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
+			})
+			if err != nil {
+				return err
+			}
+
+			for _, pod := range pods.Items {
+				isDaemonSet := false
```
Comment (Issif) [inline actionners/kubernetes/drain/drain.go:243]: it also requires the statefulsets

**Review by Issif — CHANGES_REQUESTED** [review]

File: actionners/kubernetes/drain/drain.go (hunk 3 above)
Comment (IgorEulalio) [reply actionners/kubernetes/drain/drain.go:244]: the logic here is to release the "holder" to the next action whenever there are no pods running on the node or all the pods running are daemonsets that we can ignore; why would we have StatefulSets in the logic too? We want to wait StatefulSets to be evicted before going to the next action.

File: actionners/kubernetes/drain/drain.go [hunk 4]
```diff
@@ -92,47 +100,53 @@ func Action(action *rules.Action, event *events.Event) (utils.LogLine, *model.Da
...
 				if config.IgnoreStatefulSets {
 					ignoredPodsCount++
 				}
-			case "ReplicaSet":
-				replicaSetName, err := kubernetes.GetOwnerName(p)
+			case utils.ReplicaSetStr:
+				var replicaSetName string
+				replicaSetName, err = kubernetes.GetOwnerName(p)
 				if err != nil {
 					utils.PrintLog("warning", utils.LogLine{Message: fmt.Sprintf("error getting pod owner name: %v", err)})
```
Comment (Issif) [inline actionners/kubernetes/drain/drain.go:127]:
```suggestion
					utils.PrintLog("warning", utils.LogLine{Message: "action", Output: fmt.Sprintf("error getting pod owner name: %v", err)})
```

Comment (Issif) [reply actionners/kubernetes/drain/drain.go:127]: The Message field is used to give the context (init, action, notification, ...), use Output for the real "message"

File: actionners/kubernetes/drain/drain.go [hunk 5]
```diff
@@ -92,47 +100,53 @@ func Action(action *rules.Action, event *events.Event) (utils.LogLine, *model.Da
...
 					otherErrorsCount++
 				}
 				if config.MinHealthyReplicas != "" {
-					replicaSet, err := client.GetReplicaSet(replicaSetName, p.Namespace)
+					var replicaSet *v1.ReplicaSet
```
Comment (Issif) [inline actionners/kubernetes/drain/drain.go:131]: You should start by checking if `config.MinHealthyReplicas != ""` right after the `case utils.ReplicaSetStr`.

File: actionners/kubernetes/drain/drain.go [hunk 6]
```diff
@@ -166,14 +181,25 @@ func Action(action *rules.Action, event *events.Event) (utils.LogLine, *model.Da
...
+			if err = client.PolicyV1().Evictions(pod.GetNamespace()).Evict(context.Background(), eviction); err != nil {
 				utils.PrintLog("warning", utils.LogLine{Message: fmt.Sprintf("error evicting pod '%v': %v", p.Name, err)})
 				evictionErrorsCount++
 			}
 		}()
 	}
 
 	wg.Wait()
+	if config.WaitPeriod != 0 {
+		err = verifyEvictionHasFinished(client, config.WaitPeriod, nodeName, config)
```
Comment (Issif) [inline actionners/kubernetes/drain/drain.go:193]: the client can be retrieved in the func directly, no need to pass it as arg

**Review by Issif — CHANGES_REQUESTED** [review]

File: actionners/kubernetes/drain/drain.go (hunk 6 above)
Comment (IgorEulalio) [reply actionners/kubernetes/drain/drain.go]: I think it's better to receive the client so we can later test this function by passing a mock client, the client would be refactored in a interface and then we'd use that interface as a way to mock it.

**Pull request closed without merging**